// bootstrap-vue-next/dist/BModalOrchestrator.vue_vue_type_script_setup_true_lang-BtqcHXII.mjs
import { defineComponent, createBlock, openBlock, unref, withCtx, createCommentVNode, normalizeProps, mergeProps } from "vue";
import { u as useDefaults } from "./useDefaults-DESCdb0p.mjs";
import { u as useModal } from "./index-CyEArZ56.mjs";
import { _ as _sfc_main$1 } from "./ConditionalTeleport.vue_vue_type_script_lang-Cp-egLmm.mjs";
import { _ as _sfc_main$2 } from "./BOrchestrator.vue_vue_type_script_setup_true_lang-Dz_rJ4Bq.mjs";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    inheritAttrs: false
  },
  __name: "BModalOrchestrator",
  props: {
    teleportDisabled: { type: Boolean, default: false },
    teleportTo: { default: "body" }
  },
  setup(__props, { expose: __expose }) {
    const _props = __props;
    const props = useDefaults(_props, "BModalOrchestrator");
    const tools = useModal();
    let renderOrchestrator = false;
    if (!tools._isOrchestratorInstalled.value) {
      renderOrchestrator = true;
    }
    if (process.env.NODE_ENV === "development") {
      console.warn("BModalOrchestrator is deprecated, use BOrchestrator instead");
    }
    __expose({
      ...tools
    });
    return (_ctx, _cache) => {
      return openBlock(), createBlock(_sfc_main$1, {
        to: unref(props).teleportTo,
        disabled: unref(props).teleportDisabled
      }, {
        default: withCtx(() => [
          unref(renderOrchestrator) ? (openBlock(), createBlock(_sfc_main$2, normalizeProps(mergeProps({ key: 0 }, _ctx.$attrs)), null, 16)) : createCommentVNode("", true)
        ]),
        _: 1
      }, 8, ["to", "disabled"]);
    };
  }
});
export {
  _sfc_main as _
};
