// bootstrap-vue-next/dist/keys-BLeKMItg.mjs
const genericBvnPrefix = "BootstrapVueNext__";
const withBvnPrefix = (value, suffix = "") => {
  const suffixWithTrail = `${suffix}___`;
  return `${genericBvnPrefix}ID__${value}__${suffix ? suffixWithTrail : ""}`;
};
const createBvnInjectionKey = (name) => withBvnPrefix(name);
const createBvnRegistryInjectionKey = (name) => withBvnPrefix(`${name}__registry`);
const carouselInjectionKey = createBvnInjectionKey("carousel");
const tabsInjectionKey = createBvnInjectionKey("tabs");
const progressInjectionKey = createBvnInjectionKey("progress");
const listGroupInjectionKey = createBvnInjectionKey("listGroup");
const avatarGroupInjectionKey = createBvnInjectionKey("avatarGroup");
const accordionInjectionKey = createBvnInjectionKey("accordion");
const checkboxGroupKey = createBvnInjectionKey("checkboxGroup");
const radioGroupKey = createBvnInjectionKey("radioGroup");
const collapseInjectionKey = createBvnInjectionKey("collapse");
const showHideRegistryKey = createBvnRegistryInjectionKey("showHide");
const dropdownInjectionKey = createBvnInjectionKey("dropdown");
const navbarInjectionKey = createBvnInjectionKey("navbar");
const rtlRegistryKey = createBvnRegistryInjectionKey("rtl");
const breadcrumbGlobalIndexKey = `${genericBvnPrefix}global_breadcrumb`;
const breadcrumbRegistryKey = createBvnRegistryInjectionKey("breadcrumb");
const modalManagerKey = createBvnRegistryInjectionKey("modalManager");
const defaultsKey = createBvnRegistryInjectionKey("defaults");
const inputGroupKey = createBvnInjectionKey("inputGroup");
const buttonGroupKey = createBvnInjectionKey("buttonGroup");
const orchestratorRegistryKey = createBvnRegistryInjectionKey("orchestrator");
const formGroupKey = createBvnInjectionKey("formGroupPlugin");
const formSelectKey = createBvnInjectionKey("formSelect");
const tableKeyboardNavigationKey = createBvnInjectionKey("tableKeyboardNavigation");
export {
  breadcrumbGlobalIndexKey as a,
  buttonGroupKey as b,
  collapseInjectionKey as c,
  dropdownInjectionKey as d,
  breadcrumbRegistryKey as e,
  defaultsKey as f,
  accordionInjectionKey as g,
  avatarGroupInjectionKey as h,
  inputGroupKey as i,
  checkboxGroupKey as j,
  formGroupKey as k,
  radioGroupKey as l,
  modalManagerKey as m,
  formSelectKey as n,
  orchestratorRegistryKey as o,
  listGroupInjectionKey as p,
  progressInjectionKey as q,
  rtlRegistryKey as r,
  showHideRegistryKey as s,
  tabsInjectionKey as t,
  carouselInjectionKey as u,
  navbarInjectionKey as v,
  withBvnPrefix as w,
  tableKeyboardNavigationKey as x
};
